### PR TITLE
Add custom download handler support to COM api

### PR DESCRIPTION
**Pull request by yao-msft**
This is part of the effort to support private 3rd party sources where packages hosted may require custom handling (encryption, custom format, etc).

- Added support for a group policy source to be configured as hidden
- Added custom download callback functionality
- Added e2e tests
 ###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://api.github.com/repos/microsoft/winget-cli/pulls/2912)
 ###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://api.github.com/repos/microsoft/winget-cli/pulls/2912)

### CODE REVIEW

File: src/AppInstallerCommonCore/GroupPolicy.cpp [hunk 1]
```diff
@@ -196,6 +196,14 @@ namespace AppInstaller::Settings
                 return std::nullopt;
             }
 
+            // IsHidden attribute is optional. When a source is hidden, it can only be reached with direct name.
+            const std::string isHiddenName = "IsHidden";
+            if (sourceJson.isMember(isHiddenName) &&sourceJson[isHiddenName].isBool())
```
Comment (ryfu-msft) [inline src/AppInstallerCommonCore/GroupPolicy.cpp:201]: nit:
```suggestion
            if (sourceJson.isMember(isHiddenName) && sourceJson[isHiddenName].isBool())

```

**Review by ryfu-msft — APPROVED** [review]

File: src/AppInstallerRepositoryCore/RepositorySource.cpp [hunk 2]
```diff
@@ -354,7 +354,7 @@ namespace AppInstaller::Repository
         }
         else
         {
-            auto source = sourceList.GetCurrentSource(name);
+            auto source = includeHidden ? sourceList.GetSource(name) : sourceList.GetCurrentSource(name);
```
Comment (florelis) [inline src/AppInstallerRepositoryCore/RepositorySource.cpp:357]: 1. I don't like the name of `GetSource()` and `GetCurrentSource()`, I had absolutely no idea what the difference was until I went to look at the definition.
2. The comments for these two functions mention tombstones. We should also add a mention of hidden by policy.
3. Would this resurrect tombstone-d sources?

File: src/AppInstallerRepositoryCore/Public/winget/RepositorySource.h [hunk 3]
```diff
@@ -163,7 +163,7 @@ namespace AppInstaller::Repository
         Source();
 
         // Constructor to get a named source, passing empty string will get all available sources.
-        Source(std::string_view name);
+        Source(std::string_view name, bool includeHidden = false);
```
Comment (florelis) [inline src/AppInstallerRepositoryCore/Public/winget/RepositorySource.h:166]: I may have missed this, but is there any caller that passes true here now?

File: src/AppInstallerCLICore/Commands/RootCommand.cpp [hunk 4]
```diff
@@ -36,7 +36,10 @@ namespace AppInstaller::CLI
             Execution::TableOutput<3> sourcesTable{ context.Reporter, { header, Resource::String::SourceListType, Resource::String::SourceListArg } };
             for (const auto& source : sources)
             {
-                sourcesTable.OutputLine({ source.Name, source.Type, source.Arg });
+                if (!source.IsHidden)
+                {
+                    sourcesTable.OutputLine({ source.Name, source.Type, source.Arg });
+                }
```
Comment (florelis) [inline src/AppInstallerCLICore/Commands/RootCommand.cpp:42]: Why do we want to hide the sources from the list of Group Policy sources? Is there a way to show them?
What about the list in `source list`?

File: src/AppInstallerCLICore/Workflows/DownloadFlow.h [hunk 5]
```diff
@@ -1,10 +1,17 @@
...
+#include <AppInstallerProgress.h>
+
+namespace AppInstaller::CLI::Execution
+{
+    struct Context;
+}
 
 namespace AppInstaller::CLI::Workflow
 {
+    typedef std::function<HRESULT(std::string url, std::string path, std::vector<std::uint8_t> hash, AppInstaller::IProgressCallback& progress)> PackageDownloadHandlerFunction;
```
[the remark this answers is not in the document]
Comment (yao-msft) [reply src/AppInstallerCLICore/Workflows/DownloadFlow.h:13]: I copied from another typedef std::function somewhere in our COM implementation. Sure, I'll update this one and also the one I copied from.

**Comment by yao-msft** [comment]
The feature is postponed. This pr could be reactivated in future releases when needed.

**Pull request closed without merging**